File: algorithms/8.swift
// https://leetcode.com/problems/string-to-integer-atoi/

class Solution {
    func myAtoi(str: String) -> Int {
        var s = str.characters.map { String($0) }

        // remove leading spaces
        while s.count > 0 && s[0] == " " {
            s.removeFirst()
        }

        guard s.count > 0 else { return 0 }

        // extract sign
        var positive = true
        if s[0] == "+" {
            positive = true
            s.removeFirst()
        } else if s[0] == "-" {
            positive = false
            s.removeFirst()
        }

        // converting while handling overflow
        var result = 0
        let digits = Set(["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"])
        for i in 0..<s.count {
            if !digits.contains(s[i]) { break }

            if Int32.max/10 < Int32(result) || (Int32.max/10 == Int32(result) && Int32.max%10 < Int32(s[i])) {
                return positive ? Int(Int32.max) : Int(Int32.min)
            }

            result = result*10 + Int(s[i])!
        }

        return positive ? result : -result
    }
}

File: algorithms/8.py
# https://leetcode.com/problems/string-to-integer-atoi/


class Solution(object):
    def myAtoi(self, str):
        """
        :type str: str
        :rtype: int
        """

        # extract [['sign', 'str']] using regex
        s = re.findall('(^[\+\-]{0,1})0*(\d+)', str.strip())
        if len(s) == 0:
            return 0
        positive = True if s[0][0] != '-' else False
        str = s[0][1]

        result = 0
        int_max = pow(2, 31) - 1
        int_min = -pow(2, 31)
        for i in range(len(str)):
            if ord(str[i]) < ord('0') or ord(str[i]) > ord('9'):
                break

            d = ord(str[i]) - ord('0')

            # check for overflow
            if int_max // 10 < result or (int_max // 10 == result and int_max % 10 < d):
                return int_max if positive else int_min

            result = result * 10 + d

        return result if positive else -result
